[Bug]: fix(channels): Feishu channel WebSocket crashes with cross-loop RuntimeError in multi-agent environment

## QwenPaw Version

1.0.2 (and the latest `main` branch commit)

## Description

When running multiple agents (e.g., `default` and a custom `QA_Agent`) concurrently in the same process, each configured with a different Feishu bot (`app_id`/`app_secret`), the framework crashes with a cross-loop `RuntimeError: Task attached to a different loop`. 

This occurs because the underlying Feishu Python SDK (`lark-oapi`) relies on a module-level global variable `lark_oapi.ws.client.loop` for event loop operations (like scheduling ping/pong heartbeats and task creation).

The current patching logic in `src/copaw/app/channels/feishu/channel.py` attempts to temporarily swap the global loop before awaiting the SDK's infinite message-receiving loop. However, since the loop yields control during IO, a newly started agent will overwrite the global variable with its own loop, causing all subsequent heartbeat tasks and incoming messages for the first agent to be erroneously dispatched to the second agent's thread.

**Related PR(s):** None

**Security considerations:** None

## Component(s) Affected

- [ ] Core / Backend (app, agents, config, providers, utils, local_models)
- [ ] Console (frontend web UI)
- [x] Channels (DingTalk, Feishu, QQ, Discord, iMessage, etc.)
- [ ] Skills
- [ ] CLI
- [ ] Documentation (website)
- [ ] Tests
- [ ] CI/CD
- [ ] Scripts / Deploy

## Environment

- **QwenPaw version:** 1.0.2
- **OS:** Windows
- **Install method:** from source / pip
- **Python version:** 3.12 (and likely all other supported versions)

## Steps to Reproduce

1. Configure `config.json` and agent workspaces to spawn at least two active agents in the same QwenPaw instance.
2. Provide a distinct Feishu channel configuration (different `app_id` and `app_secret`) for each agent.
3. Start the QwenPaw background process.
4. Send a message to the Feishu bot bound to the agent that started *first*.
5. The console will throw a `RuntimeError` originating from `lark_oapi/ws/client.py`.

## Actual vs Expected

- **Actual:** The first agent started never receives the message. Instead, the Feishu SDK attempts to dispatch the network payload/heartbeat to the event loop of the agent that started *last*, causing the app to crash with: `RuntimeError: Task attached to a different loop`.
- **Expected:** Multiple Feishu bots tied to different agents should be able to receive messages concurrently within their respective isolated event loops without polluting the global `lark_oapi.ws.client.loop` variable.

## Logs / Screenshots

```text
RuntimeError: Task <Task pending name='Task-5' coro=<Client._ping_loop() running at .../lark_oapi/ws/client.py>> attached to a different loop
Additional Notes
Proposed Solution:
Instead of statically replacing the global variable, we should inject a dynamic Proxy object that resolves the loop at runtime based on the calling thread.
I propose replacing the current patching logic in _create_ws_client with the following _EventLoopProxy:
        async def _patched_connect() -> None:

            import lark_oapi.ws.client as _ws_mod



            class _EventLoopProxy:

                def __getattr__(self, name):

                    try:

                        return getattr(asyncio.get_running_loop(), name)

                    except RuntimeError:

                        return getattr(asyncio.get_event_loop(), name)



            if not isinstance(getattr(_ws_mod, "loop", None), _EventLoopProxy):

                _ws_mod.loop = _EventLoopProxy()

            try:

                await _original_connect()

            finally:

                pass



        async def _patched_receive_message_loop() -> None:

            import lark_oapi.ws.client as _ws_mod



            class _EventLoopProxy:

                def __getattr__(self, name):

                    try:

                        return getattr(asyncio.get_running_loop(), name)

                    except RuntimeError:

                        return getattr(asyncio.get_event_loop(), name)



            if not isinstance(getattr(_ws_mod, "loop", None), _EventLoopProxy):

                _ws_mod.loop = _EventLoopProxy()

            try:

                await _original_recv_loop()

            finally:

                pass

With this proxy in place, whenever the Feishu SDK attempts to call loop.create_task(), the proxy dynamically returns the asyncio event loop belonging to the specific thread that triggered the call. This safely isolates multiple agents without breaking the SDK's internal logic.

## Title
[Bug] SDK WebSocket module-level `loop` modification (`saved, _ws_mod.loop = _ws_mod.loop, asyncio.get_running_loop()`) causes cross-thread event loop corruption and connection hang ("fake dead")

## Environment
- Python version: 3.12+ (also affects 3.11)
- qwenpaw version: 1.1.0
- OS: Windows

## Description
When running the Feishu/Lark WebSocket client in a multi-agent or multi-threaded environment, the recent fix that attempts to patch the `loop` issue by temporarily swapping `lark_oapi.ws.client.loop` causes severe cross-thread event loop corruption. 

After running successfully for a few hours, the WebSocket connection silently hangs (fake dead). It receives no new messages, triggers no exceptions to the main application, and fails to reconnect.

### The Root Cause
The `lark_oapi/ws/client.py` relies on a module-level variable `loop` (typically initialized via `asyncio.get_event_loop()`). 
The recent fix tries to solve cross-loop issues by doing this inside methods like `_connect` and `_receive_message_loop`:
```python
saved, _ws_mod.loop = _ws_mod.loop, asyncio.get_running_loop()
try:
    await _original_connect()
finally:
    _ws_mod.loop = saved
```

**Why this is dangerous in a background thread:**
1. When `_connect()` runs in a background thread, it temporarily sets `_ws_mod.loop` to the background thread's running loop.
2. When the `try` block finishes, `finally` executes, and `_ws_mod.loop` is **restored to `saved` (which is often the Main Thread's loop)**.
3. The WebSocket connection is still alive in the background thread. When a new message arrives hours later, the background thread executes `loop.create_task(self._handle_message(msg))`.
4. Because the loop was restored to the Main Thread's loop, the background thread is now calling `create_task()` directly on the Main Thread's loop. 
5. In Python's `asyncio`, calling `create_task()` from a different thread without using `asyncio.run_coroutine_threadsafe()` is **not thread-safe**. This causes the internal asyncio queues to corrupt, resulting in a silent deadlock (the bot stops responding entirely without crashing).

Furthermore, when the connection drops, the `_ping_loop`'s `finally: await asyncio.sleep(120)` blocks the graceful shutdown of the old connection, exacerbating the hang during reconnection attempts.

## Reproduction Steps
1. Run the `lark_oapi.ws.Client` in a dedicated background `threading.Thread` with its own `asyncio.new_event_loop()`.
2. Connect successfully and interact with the bot.
3. Wait for several hours, or simulate a network drop/reconnect scenario.
4. Observe that the bot stops receiving messages silently. The logs might show `receive message loop exit, err: no close frame received or sent` followed by a silent hang where `_reconnect()` never successfully restores message processing.

## Suggested Fix (`_EventLoopProxy` pattern)
Instead of temporarily modifying and restoring the module-level `loop` (which is inherently thread-unsafe for long-lived background tasks), the SDK should use a dynamic proxy or rely directly on `asyncio.get_running_loop()` at the exact moment a task is created.

We have successfully patched this in our downstream project (QwenPaw/AgentClaw) by replacing the module-level `loop` entirely with an `_EventLoopProxy`:

```python
class _EventLoopProxy:
    def __getattr__(self, name: str):
        try:
            # Always get the loop running in the CURRENT thread dynamically
            loop = asyncio.get_running_loop()
        except RuntimeError:
            loop = asyncio.get_event_loop()
        return getattr(loop, name)

import lark_oapi.ws.client as _ws_mod
_ws_mod.loop = _EventLoopProxy()
```

### Additional Recommendation for SDK
In `lark_oapi/ws/client.py`, the `_ping_loop` contains a `finally` block that sleeps for 120 seconds (`self._ping_interval`). If `_conn.recv()` raises a `ConnectionClosedException`, the cleanup routines often block waiting for `_ping_loop` to finish its 120s sleep. Wrapping these cleanup routines with `asyncio.wait_for(..., timeout=2.0)` or explicitly cancelling the `_ping_loop` task during `_disconnect` would drastically improve reconnection speed and reliability.

## Conclusion
Please consider removing the temporary `saved, loop = loop, current_loop` swaps in favor of dynamically resolving the event loop at runtime (e.g., calling `asyncio.get_running_loop().create_task(...)` directly in `_receive_message_loop`), as this guarantees thread safety across any architecture.

## Logs
```shell
[Lark] [2026-04-14 05:25:53,507] [ERROR] receive message loop exit, err: no close frame received or sent [conn_id=7628208161980599252]
ERROR C:\Users\Administrator\AppData\Local\Programs\Python\Python312\Lib\site-packages\lark_oapi\ws\client.py:173 | 2026-04-14 05:25:53 | receive message loop exit, err: no close frame received or sent [conn_id=7628208161980599252]
[Lark] [2026-04-14 05:25:53,507] [INFO] disconnected to wss://[msg-frontier.feishu.cn/ws/v2?fpid=493&aid=552564&device_id=7628208161980599252&access_key=4ea2eeb254dda3e977384a95fc4c7c9c&service_id=33554678&ticket=0a4268ae-012b-4cd4-bd27-dfe8255f5b5f](http://msg-frontier.feishu.cn/ws/v2?fpid=493&aid=552564&device_id=7628208161980599252&access_key=4ea2eeb254dda3e977384a95fc4c7c9c&service_id=33554678&ticket=0a4268ae-012b-4cd4-bd27-dfe8255f5b5f) [conn_id=7628208161980599252]
WARNING src\qwenpaw\app\channels\feishu\channel.py:1851 | 2026-04-14 05:25:59 | feishu WebSocket conn lost, forcing reconnect...
INFO src\qwenpaw\app\channels\feishu\channel.py:1892 | 2026-04-14 05:25:59 | feishu WebSocket event loop stopped, will attempt to reconnect
```

> ## Title
> 
> [Bug] SDK WebSocket module-level `loop` modification (`saved, _ws_mod.loop = _ws_mod.loop, asyncio.get_running_loop()`) causes cross-thread event loop corruption and connection hang ("fake dead")
> 
> ## Environment
> 
> * Python version: 3.12+ (also affects 3.11)
> * qwenpaw version: 1.1.0
> * OS: Windows
> 
> ## Description
> 
> When running the Feishu/Lark WebSocket client in a multi-agent or multi-threaded environment, the recent fix that attempts to patch the `loop` issue by temporarily swapping `lark_oapi.ws.client.loop` causes severe cross-thread event loop corruption.
> 
> After running successfully for a few hours, the WebSocket connection silently hangs (fake dead). It receives no new messages, triggers no exceptions to the main application, and fails to reconnect.
> 
> ### The Root Cause
> 
> The `lark_oapi/ws/client.py` relies on a module-level variable `loop` (typically initialized via `asyncio.get_event_loop()`). The recent fix tries to solve cross-loop issues by doing this inside methods like `_connect` and `_receive_message_loop`:
> 
> saved, _ws_mod.loop = _ws_mod.loop, asyncio.get_running_loop()
> try:
>     await _original_connect()
> finally:
>     _ws_mod.loop = saved
> **Why this is dangerous in a background thread:**
> 
> 1. When `_connect()` runs in a background thread, it temporarily sets `_ws_mod.loop` to the background thread's running loop.
> 2. When the `try` block finishes, `finally` executes, and `_ws_mod.loop` is **restored to `saved` (which is often the Main Thread's loop)**.
> 3. The WebSocket connection is still alive in the background thread. When a new message arrives hours later, the background thread executes `loop.create_task(self._handle_message(msg))`.
> 4. Because the loop was restored to the Main Thread's loop, the background thread is now calling `create_task()` directly on the Main Thread's loop.
> 5. In Python's `asyncio`, calling `create_task()` from a different thread without using `asyncio.run_coroutine_threadsafe()` is **not thread-safe**. This causes the internal asyncio queues to corrupt, resulting in a silent deadlock (the bot stops responding entirely without crashing).
> 
> Furthermore, when the connection drops, the `_ping_loop`'s `finally: await asyncio.sleep(120)` blocks the graceful shutdown of the old connection, exacerbating the hang during reconnection attempts.
> 
> ## Reproduction Steps
> 
> 1. Run the `lark_oapi.ws.Client` in a dedicated background `threading.Thread` with its own `asyncio.new_event_loop()`.
> 2. Connect successfully and interact with the bot.
> 3. Wait for several hours, or simulate a network drop/reconnect scenario.
> 4. Observe that the bot stops receiving messages silently. The logs might show `receive message loop exit, err: no close frame received or sent` followed by a silent hang where `_reconnect()` never successfully restores message processing.
> 
> ## Suggested Fix (`_EventLoopProxy` pattern)
> 
> Instead of temporarily modifying and restoring the module-level `loop` (which is inherently thread-unsafe for long-lived background tasks), the SDK should use a dynamic proxy or rely directly on `asyncio.get_running_loop()` at the exact moment a task is created.
> 
> We have successfully patched this in our downstream project (QwenPaw/AgentClaw) by replacing the module-level `loop` entirely with an `_EventLoopProxy`:
> 
> class _EventLoopProxy:
>     def __getattr__(self, name: str):
>         try:
>             # Always get the loop running in the CURRENT thread dynamically
>             loop = asyncio.get_running_loop()
>         except RuntimeError:
>             loop = asyncio.get_event_loop()
>         return getattr(loop, name)
> 
> import lark_oapi.ws.client as _ws_mod
> _ws_mod.loop = _EventLoopProxy()
> ### Additional Recommendation for SDK
> 
> In `lark_oapi/ws/client.py`, the `_ping_loop` contains a `finally` block that sleeps for 120 seconds (`self._ping_interval`). If `_conn.recv()` raises a `ConnectionClosedException`, the cleanup routines often block waiting for `_ping_loop` to finish its 120s sleep. Wrapping these cleanup routines with `asyncio.wait_for(..., timeout=2.0)` or explicitly cancelling the `_ping_loop` task during `_disconnect` would drastically improve reconnection speed and reliability.
> 
> ## Conclusion
> 
> Please consider removing the temporary `saved, loop = loop, current_loop` swaps in favor of dynamically resolving the event loop at runtime (e.g., calling `asyncio.get_running_loop().create_task(...)` directly in `_receive_message_loop`), as this guarantees thread safety across any architecture.
> 
> ## Logs
> 
> [Lark] [2026-04-14 05:25:53,507] [ERROR] receive message loop exit, err: no close frame received or sent [conn_id=7628208161980599252]
> ERROR C:\Users\Administrator\AppData\Local\Programs\Python\Python312\Lib\site-packages\lark_oapi\ws\client.py:173 | 2026-04-14 05:25:53 | receive message loop exit, err: no close frame received or sent [conn_id=7628208161980599252]
> [Lark] [2026-04-14 05:25:53,507] [INFO] disconnected to wss://[msg-frontier.feishu.cn/ws/v2?fpid=493&aid=552564&device_id=7628208161980599252&access_key=4ea2eeb254dda3e977384a95fc4c7c9c&service_id=33554678&ticket=0a4268ae-012b-4cd4-bd27-dfe8255f5b5f](http://msg-frontier.feishu.cn/ws/v2?fpid=493&aid=552564&device_id=7628208161980599252&access_key=4ea2eeb254dda3e977384a95fc4c7c9c&service_id=33554678&ticket=0a4268ae-012b-4cd4-bd27-dfe8255f5b5f) [conn_id=7628208161980599252]
> WARNING src\qwenpaw\app\channels\feishu\channel.py:1851 | 2026-04-14 05:25:59 | feishu WebSocket conn lost, forcing reconnect...
> INFO src\qwenpaw\app\channels\feishu\channel.py:1892 | 2026-04-14 05:25:59 | feishu WebSocket event loop stopped, will attempt to reconnect

Hi @JohnYWN, thank you for the detailed analysis and the `_EventLoopProxy` suggestion!

I've adopted your proposed approach. The fix replaces the static `saved, _ws_mod.loop = _ws_mod.loop, asyncio.get_running_loop()` / `finally: restore` pattern with a module-level `_EventLoopProxy` that dynamically resolves `asyncio.get_running_loop()` on every attribute access. The proxy is installed once at import time, so there's no need for per-method monkey-patching or restoration — this eliminates the stale-loop problem across reconnect cycles.

Regarding the `_ping_loop` 120s sleep concern: I've analyzed the interaction between the SDK's `_ping_loop` and our `_run_ws_forever` architecture. In our implementation, when a disconnection is detected by `_monitor_connection_health` (which polls every 30s), it calls `self._ws_loop.stop()`, and the `finally` block in `_run_ws_forever` explicitly cancels all pending tasks (including `_ping_loop`) via `task.cancel()` before closing the loop. This means the 120s sleep in `_ping_loop`'s `finally` block is forcefully interrupted and does not block our reconnection — as your own logs confirm (reconnect triggered ~6s after disconnection, not 120s). So this is not an issue under our current architecture.

The fix will be included in the upcoming PR. Thanks again for your contribution!

--- a/src/qwenpaw/app/channels/feishu/channel.py
+++ b/src/qwenpaw/app/channels/feishu/channel.py
@@ -101,6 +101,17 @@ else:
         )
         _declare_namespace_patched = True
 
+
+class _EventLoopProxy:
+    """Resolve ``lark_oapi.ws.client.loop`` to the calling thread's loop."""
+
+    def __getattr__(self, name: str) -> Any:
+        try:
+            return getattr(asyncio.get_running_loop(), name)
+        except RuntimeError:
+            return getattr(asyncio.get_event_loop(), name)
+
+
 try:
     import lark_oapi as lark
     from lark_oapi.api.contact.v3 import GetUserRequest
@@ -118,6 +129,10 @@ try:
         GetMessageResourceRequest,
         P2ImMessageReceiveV1,
     )
+
+    import lark_oapi.ws.client as _ws_mod
+
+    _ws_mod.loop = _EventLoopProxy()
 except ImportError:  # pragma: no cover - optional dependency may be missing
     lark = None  # type: ignore[assignment]
     GetUserRequest = None  # type: ignore[assignment]
@@ -1966,68 +1981,8 @@ class FeishuChannel(BaseChannel):
                 receive_id_type,
             )
 
-    def _create_ws_client(self) -> Any:
-        """Create a fresh ``lark.ws.Client`` with loop-isolated internals.
-
-        The SDK uses a module-level ``loop`` for ``create_task`` calls,
-        which causes cross-loop errors when multiple agents run in
-        separate threads.  We patch ``_connect`` and
-        ``_receive_message_loop`` to swap in the running loop before
-        delegating to the originals.
-        """
-        event_handler = (
-            lark.EventDispatcherHandler.builder(
-                self.encrypt_key,
-                self.verification_token,
-            )
-            .register_p2_im_message_receive_v1(self._on_message_sync)
-            .build()
-        )
-        client = lark.ws.Client(
-            self.app_id,
-            self.app_secret,
-            event_handler=event_handler,
-            log_level=lark.LogLevel.INFO,
-            domain=(
-                lark.LARK_DOMAIN
-                if self.domain == "lark"
-                else lark.FEISHU_DOMAIN
-            ),
-        )
-
-        _original_connect = client._connect
-        _original_recv_loop = client._receive_message_loop
-
-        async def _patched_connect() -> None:
-            import lark_oapi.ws.client as _ws_mod
-
-            saved, _ws_mod.loop = _ws_mod.loop, asyncio.get_running_loop()
-            try:
-                await _original_connect()
-            finally:
-                _ws_mod.loop = saved
-
-        async def _patched_receive_message_loop() -> None:
-            import lark_oapi.ws.client as _ws_mod
-
-            saved, _ws_mod.loop = _ws_mod.loop, asyncio.get_running_loop()
-            try:
-                await _original_recv_loop()
-            finally:
-                _ws_mod.loop = saved
-
-        client._connect = _patched_connect
-        client._receive_message_loop = _patched_receive_message_loop
-        return client
-
     def _run_ws_forever(self) -> None:
-        """Run WebSocket with automatic reconnection (exponential backoff).
-
-        Each iteration creates a fresh event loop and ``ws.Client``, then
-        drives the connection directly (``_connect`` → ``_ping_loop`` →
-        ``_select``) instead of calling the SDK's ``start()`` which relies
-        on a shared module-level ``loop``.
-        """
+        """Run WebSocket with exponential-backoff reconnection."""
         retry_delay = FEISHU_WS_INITIAL_RETRY_DELAY
 
         while not self._stop_event.is_set() and not self._closed:
@@ -2035,7 +1990,27 @@ class FeishuChannel(BaseChannel):
             asyncio.set_event_loop(self._ws_loop)
             connection_started = False
             try:
-                self._ws_client = self._create_ws_client()
+                event_handler = (
+                    lark.EventDispatcherHandler.builder(
+                        self.encrypt_key,
+                        self.verification_token,
+                    )
+                    .register_p2_im_message_receive_v1(
+                        self._on_message_sync,
+                    )
+                    .build()
+                )
+                self._ws_client = lark.ws.Client(
+                    self.app_id,
+                    self.app_secret,
+                    event_handler=event_handler,
+                    log_level=lark.LogLevel.INFO,
+                    domain=(
+                        lark.LARK_DOMAIN
+                        if self.domain == "lark"
+                        else lark.FEISHU_DOMAIN
+                    ),
+                )
 
                 async def _select() -> None:
                     while True:

